client/db/bolt/db.go

```diff
@@ -1624,6 +1624,167 @@ func (idx *timeIndexNewest) add(t uint64, k []byte, b *bbolt.Bucket) {
+func (db *BoltDB) DeleteInactiveOrders(ctx context.Context, olderThan *time.Time, perBatchFn func(ords []*db.MetaOrder) error) error {
+	start := time.Now()
+	const batchSize = 1000
+	var (
+		finished   bool
+		olderThanB []byte
+	)
+	if olderThan != nil {
+		olderThanB = uint64Bytes(encode.UnixMilliU(*olderThan))
+	} else {
+		olderThanB = uint64Bytes(encode.UnixMilliU(start))
+	}
+	for !finished {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		deletedOrders := make([]*dexdb.MetaOrder, 0, batchSize)
+		err := db.Update(func(tx *bbolt.Tx) error {
+			// Some archived orders may still be needed for active matches.
+			// Put those order id's in a map to prevent deletion.
+			amb := tx.Bucket(activeMatchesBucket)
+			activeMatches := make(map[order.OrderID]struct{})
+			if err := amb.ForEach(func(k, _ []byte) error {
+				mBkt := amb.Bucket(k)
+				if mBkt == nil {
+					return fmt.Errorf("match %x bucket is not a bucket", k)
+				}
+
+				oidB := mBkt.Get(orderIDKey)
+				var oid order.OrderID
+				copy(oid[:], oidB)
+				activeMatches[oid] = struct{}{}
+				return nil
+			}); err != nil {
+				return fmt.Errorf("unable to get active matches: %v", err)
+			}
+
+			// Run through the archived order bucket storing id's keys for
+			// returning the order data and deletion later.
+			archivedOB := tx.Bucket(archivedOrdersBucket)
+			if archivedOB == nil {
+				return fmt.Errorf("failed to open %s bucket", string(archivedOrdersBucket))
+			}
+			// Retrieve one more than we will delete in order to guage if
+			// this is the last order to delete.
+			oneOverSize := batchSize + 1
+			filter := func(k []byte, oBkt *bbolt.Bucket) bool {
+				var oid order.OrderID
+				copy(oid[:], k)
+				if order.OrderStatus(intCoder.Uint16(oBkt.Get(statusKey))).IsActive() {
+					db.log.Warnf("active order %v found in inactive bucket", oid)
+					return false
+				}
+				if _, has := activeMatches[oid]; has {
+					return false
+				}
+				timeB := oBkt.Get(updateTimeKey)
+				return bytes.Compare(timeB, olderThanB) <= 0
+			}
+			trios := newestBuckets([]*bbolt.Bucket{archivedOB}, oneOverSize, updateTimeKey, filter)
+
+			// Ignore the last order if it exists. Otherwise there are no
+			// more orders to delete.
+			if len(trios) == oneOverSize {
+				trios = trios[:batchSize]
+			} else {
+				finished = true
+			}
+			for _, trio := range trios {
+				o, err := decodeOrderBucket(trio.k, trio.b)
+				if err != nil {
+					return fmt.Errorf("failed to decode order bucket: %v", err)
+				}
+				deletedOrders = append(deletedOrders, o)
+				if err := archivedOB.DeleteBucket(trio.k); err != nil {
+					return fmt.Errorf("failed to delete order bucket: %v", err)
+				}
+			}
+			db.log.Infof("Deleted %d orders from the database in %v.", len(trios), time.Since(start))
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("unable to delete orders: %v", err)
+		}
+		if perBatchFn != nil {
+			if err := perBatchFn(deletedOrders); err != nil {
```

Comment:
I had envisioned calling a `func(ords *db.MetaOrder) error` for each order _within_ the transaction (one at a time, no slice) so that it can be rolled back if processing the removed order(s) failed, and they would be more continuously streaming.  Down here it's too late to get those orders back in the DB without manually reinserting them, and `perBatchFn` failed to process them so they didn't make it to the consumer.

Do you think it would be problematic or less performant to have the function take a single order and to stream them continuously from within the db transaction that loads and removes them?  I haven't considered this as much as you so I could be mistaken.

Reply:
imo the performance isn't such a big issue because this will not be used regularly, or by everyone. Will make this change so that we can roll back the tx if the passed function fails.

If we still do it inside the batch, however, we will return some orders a deleted that are later un-deleted with the roll back though. Will log this in case.

Reply:
Those are good points.  Only make the changes if they feel right to you.

Reply:
Doing per tx and logging the number of things that are rolled back if there is a problem.